File: Array/maxmin.py

from cmath import inf


def getminmax(array, size):
    max = -inf
    min = inf
    for i in range(size):
        if array[i] > max:
            max = array[i]
        if array[i] < min:
            min = array[i]
    return max, min

# APPROACH 2 - divide and conquer
def divide(array, start, end):
    size = end-start + 1
    #sprint(size)
    min = max = 0
    if size == 0 or start > end:
        return min, max
    if size == 1:
        max = min = array[start]
    elif size == 2:
        if array[start]< array[end]:
            min = array[start]
            max = array[end]
        else:
            max = array[start]
            min = array[end]
    else:
        min1, max1 = divide(array, start, start + (end-start)//2)
        min2, max2 = divide(array, start + (end-start)//2 + 1, end)
        min = min1 if min1 < min2 else min2
        max = max1 if max1 > max2 else max2
    #print(min, max)
    return min, max


array = [111,2,320,40,5,6]
size = 6
#max, min= getminmax(array, size)
max, min= divide(array, 0, size-1)

print(max, min)

File: Array/maxmin.cpp
#include<iostream>

class minmax{
public:
    int max = 0;
    int min = 0;
};

minmax getminmax(int array[], int size){
    minmax mm = minmax();
    mm.max = mm.min = array[0];
    for(int i=1; i< size; i++){
        if (array[i] < mm.min)
            mm.min = array[i];
        else if (array[i] > mm.max)
            mm.max = array[i];
    }
    return mm;
}

minmax solve(int array[], int size){
    minmax mm = minmax();
    int i=0;
    // even
    if (size%2 == 0){
        if (array[0] < array[1]){
            mm.min = array[0];
            mm.max = array[1];
        }
        else{
            mm.max = array[0];
            mm.min = array[1];
        } 
        i = 2;
    }
    else{
        //odd
        mm.max = mm.min = array[0];
        i = 1;
    }
    for(i; i< size; i+=2){
        int smaller, larger;
        if (array[i] < array[i+1]){
            larger = array[i+1];
            smaller = array[i];
        }
        else{
            larger = array[i];
            smaller = array[i+1];
        }
        if (mm.max < larger)
            mm.max = larger;
        if (mm.min > smaller)
            mm.min= smaller;
    }
    return mm;
}

int main(){
    int array[] = {111,2,32,40,5,6};
    //minmax m = getminmax(array, 6);
    minmax m = solve(array, 6);

    std::cout<<m.max<<"\n"<<m.min;
    return 0;
}